compiler/rustc_resolve/src/diagnostics.rs

```diff
@@ -1634,6 +1642,33 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             struct_span_err!(self.tcx.sess, ident.span, E0603, "{} `{}` is private", descr, ident);
         err.span_label(ident.span, format!("private {}", descr));
 
+        if let Some((this_res, outer_ident)) = outermost_res {
+            let import_suggestions = self.lookup_import_candidates(
+                outer_ident,
+                Namespace::TypeNS,
+                &parent_scope,
+                &|res: Res| res == this_res,
+            );
+            let point_to_def = !show_candidates(
+                self.tcx,
+                &mut err,
+                Some(dedup_span),
+                &import_suggestions,
+                Instead::Yes,
+                FoundUse::Yes,
+                DiagnosticMode::Import,
+                vec![],
+                "",
+            );
+            // If we suggest importing a public re-export, don't point at the definition.
+            if point_to_def && ident.span != outer_ident.span {
```

Comment:
```suggestion
            if point_to_def {
```
Not clear what the second condition came from, what case does it cover?

Reply:
Avoids

```
LL | impl a::B for S {}
   |         ^
   |         |
   |         private trait
   |         trait `B` is not publicly reachable
```

The effects aren't seen because it is the lack of changes to the existing tests (and when splitting the commits this condition wasn't something that I mentally flagged as noteworthy when cherrypicking changes).